### ISSUE TITLE
Replace groovy build-tools (buildSrc) with java

### ISSUE THREAD

**Issue by atorok** [elastic/elasticsearch#34459]
We want to have build code implemented in Java to make it easier to maintain. 
We can easily mix java and groovy implementations and everything is already set up to do so, so it's already easy to create new classes in Java. 

Here's a list of what is still in Groovy: 

-  stand alone tasks and plugins 
 - [ ] ConcatFilesTask
 - [ ] TestWithDependenciesPlugin
 - [ ] NoticeTask
 - [ ] PluginPropertiesTask
 - [ ] PluginPropertiesExtension
 - [ ] MavenFilteringHack
 - [ ] FileContentsTask
 - [ ] DependenciesInfoTask
 - [ ] VersionCollection
 - [ ] MessyTestPlugin ( possibly dead code elastic/x-plugins/issues/724 )
 - [ ] EmptyDirTask
- [ ] precommit
 - FilePermissionsTask
 - ForbiddenPatternsTask
 - UpdateShasTask
 - DependencyLicensesTask
 - PrecommitTasks
- [ ] doc
 - RestTestFromSnippetsTaskTests
 - RestTestsFromSnippetsTask
 - DocsTestPlugin 
 - SnippetsTask
 - RestTestsFromSnippetsTask
- [ ] ant and and fixtures 
 - LicenseHeadersTask
 - AntTask
-  [ ] dependent on the new testclusters implementation
 - AntFixture ( used in multipe build scripts )
 - RestTestPlugin
 - StandaloneRestTestPlugin
 - Fixture
 - RestIntegTestTask
 - RunTask
 - ClusterConfiguration
 - NodeInfo
 - ClusterFormationTasks
- [ ] vagrant 
 - VagrantFixture
 - VagrantLoggerOutputStream
 - VagrantCommandTask
 - VagrantTestPlugin
 - BatsOverVagrantTask
 - TapLoggerOutputStream
 - VagrantPropertiesExtension
- plugins 
 - [ ] BuildPlugin
 - [ ] StandaloneTestPlugin
 - [ ] PluginBuildPlugin
 - [ ] RestTestPlugin
 - [ ] StandaloneRestTestPlugin
 - [ ] VagrantSupportPlugin
- [ ] carrotsearch : either push changes upstream or replace with Gradle runner ( #31496 )

**Comment by pt429sy** [comment]
Hello, I would like to work on this.

**Comment by baptistemesta** [comment]
just for info, I am working on `NoticeTask`

**Comment by Megamiun** [comment]
Doing UpdateShasTask, probably will make a PR tonight.

**Comment by fabiodrg666** [comment]
I am interested in giving a try to `PluginPropertiesExtension`. From my understanding it should be quite straightforward.

**Comment by cdschneider** [comment]
Going to give `ForbiddenPatternsTask` a shot, FYI

**Comment by GuptaNishant** [comment]
I will like to work on DependenciesInfoTask

**Comment by akki** [comment]
I tried porting [vagrant.VagrantPropertiesExtension](https://github.com/elastic/elasticsearch/compare/master...akki:34459-vagrant?expand=1) file to Java but it seems all modules inside `vagrant` package have to be ported in one go to make it work; Please let me know if this work on `VagrantPropertiesExtension` can anyhow be used/merged separately, otherwise anybody who aims to port all the other files can feel free to use this patch.

@atorok From what I understand, https://github.com/elastic/elasticsearch/pull/34769 can be added against `NoticeTask` in the first comment.

**Comment by paras885** [comment]
Going to give `MavenFilteringHack` a shot, FYI

**Comment by akki** [comment]
Due to long inactivity on #34476 I am picking up `ConcatFilesTask`.

**Comment by bsamartins** [comment]
@atorok, are there any updates on merges / pieces that have merged?

I can take the PrecommitTasks

**Comment by atorok** [comment]
@bsamartins you will need #39605 for that

**Comment by atorok** [comment]
@bsamartins you will need the other PRs merge first as you won't be able to reference Groovy classes in Java code.

**Comment by rhamedy** [comment]
Is this in the PR description `PluginBuildPlugin #39605` correct? Shouldn't the PR #39605 be added besides `PluginPropertiesTask & PluginPropertiesExtension`?

**Comment by dRadest** [comment]
I'd like to give DependenciesInfoTask a try if nobody's working on it

**Comment by michal-rozak** [comment]
Hi @atorok, I would like to contribute but I can see that tasks above are not updated. Which task can be taken next?

**Comment by atorok** [comment]
Thanks for your interest @michal-rozak . I Updated the task to mark the ones that were completed.

**Comment by Megamiun** [comment]
@atorok I think the correct PR about the DependencyLicensesTask and UpdateShasTask is the  #41921 one. The one listed is the old one

**Comment by michal-rozak** [comment]
@atorok, I want to pick PrecommitTasks, ok?

**Comment by michal-rozak** [comment]
@atorok Oh, it seems that PrecommitTasks is taken/in progress. Maybe DocsTestPlugin then?

**Comment by atorok** [comment]
I don't think @bsamartins did any work on PrecommitTasks so I think you could take either one @michal-rozak

**Comment by jbonn360** [comment]
Hi @atorok, I'd like to work on MavenFilteringHack if no one's working on it

**Comment by rh-id** [comment]
hello @alpar-t  would like to work on `AntTask` if no one's working on it

**Comment by alpar-t** [comment]
@mark-vieira where do we stand on the `AntTask` ?

**Comment by mark-vieira** [comment]
My preference would be to remove it, and refactor the two remaining subclasses to not rely on Ant.

**Comment by mrmodise** [comment]
@mark-vieira @alpar-t I just pulled the latest code and I see `PrecommitTasks` is still pending migration. I would like to take it on if its not hanging in the PR state. Thanks

**Comment by sachinpaliwal1981** [comment]
is there any class pending from the list where work is not in progress already and can be picked up by me.

**Comment by mark-vieira** [comment]
@sachinpaliwal1981 Both `MavenFIlteringHack` and `NoticeTask` are good candiates that have no dependencies on other Groovy code and are mostly a strait conversion to Java.

**Comment by MarkRibeiro** [comment]
I would like to work on this, where can i start?

**Comment by edh-oss** [comment]
I'd like to work on the code related to docs, SnippetTask and the related classes.

I have a couple questions:

I've seen both junit and spock used for tests.  Is there one test framework that you want to make the standard, or does it not matter?

Is there any documentation that spells out things like "This is what constitutes a valid code snippet", or is the existing implementation effectively the documentation as well?

**Comment by pugnascotia** [comment]
@breskeby am I correct in thinking that this is something that you've been working on?

**Comment by breskeby** [comment]
@pugnascotia yes there is not much groovy left these days but the stuff that is left is more than just a simple port

**Comment by morantaf** [comment]
Hey, I'm going to give a go to the doc part
